platform/barefoot/sonic-platform-modules-ufispace/s9180-32x/sonic_platform/fan.py

```diff
@@ -0,0 +1,177 @@
+            device = "PSU{}".format(self.fans_psu_index)
+        else:
+            if self.fan_index == 1:
+                pos = "f"
+            else:
+                pos = "r"
+            attr = "fan{}_{}_speed_rpm".format(self.fantray_index, pos)
+            device = "FAN-CTRL"
+
+        output = self.pddf_obj.get_attr_name_output(device, attr)
+
+        if output is None:
+            return rpm_speed
+
+        output['status'] = output['status'].rstrip()
+        if output['status'].isalpha():
+            return rpm_speed
+        else:
+            rpm_speed = int(float(output['status']))
+
+        return rpm_speed
+
+    def get_direction(self):
+        """
+        Retrieves the direction of fan
+        Returns:
+            A string, either FAN_DIRECTION_INTAKE or FAN_DIRECTION_EXHAUST
+            depending on fan direction
+        """
+        direction = self.FAN_DIRECTION_INTAKE
+        if self.is_psu_fan:
+            attr = "psu_fan{}_dir".format(self.fan_index)
+            device = "PSU{}".format(self.fans_psu_index)
+        else:
+            attr = "fan{}_dir".format(self.fantray_index)
+            device = "FAN-CTRL"
+
+        output = self.pddf_obj.get_attr_name_output(device, attr)
+        if not output:
+            return direction
+
+        mode = output['mode']
+        val = output['status'].strip()
+        vmap = self.plugin_data['FAN']['direction'][mode]['valmap']
+
+        if val in vmap:
+            direction = vmap[val]
+
+        return direction
+
+    def get_presence(self):
+        """
+        Retrieves the presence of the device
+        Returns:
+            bool: True if device is present, False if not
+        """
+        presence = False
+        if self.is_psu_fan:
+            attr = "psu_present"
+            device = "PSU{}".format(self.fans_psu_index)
+        else:
+            attr = "fan{}_present".format(self.fantray_index)
+            device = "FAN-CTRL"
+
+        output = self.pddf_obj.get_attr_name_output(device, attr)
+        if not output:
+            return presence
+
+
+        mode = output['mode']
+        val = output['status'].strip()
+        vmap = self.plugin_data['FAN']['present'][mode]['valmap']
+
+        if val in vmap:
+            presence = vmap[val]
+
+        return presence
+
+    def get_status(self):
+        """
+        Retrieves the operational status of the device
+
+        Returns:
+            A boolean value, True if device is operating properly, False if not
+        """
+        speed = self.get_speed_rpm()
+        status = True if (speed != 0) else False
+        return status
+
+    def get_target_speed(self):
+        """
+        Retrieves the target (expected) speed of the fan
+        Returns:
+            An integer, the percentage of full fan speed, in the range 0 (off)
+                 to 100 (full speed)
+        """
+        if self.is_psu_fan:
+            return 75
+        else:
+            return 65
```

Comment:
Is your system fan doesnt support speed variations etc?
How is the target speed has a fixed value? This is supposed to be the speed intended by the user. Its calculation is based on the duty-cycle/pwm calculations.

Reply:
Yes, The fan speed is fully controlled by BMC and not open for the users. Still I will update the code to get current speed to avoid fixed value which is quite misleading. Thanks.